### PR TITLE
Change the Normalization algorithm

### PR DESCRIPTION
**Pull request by johnynek**
This adds a test for a bug we observed in how we
rewrite the Normalized graph. The old algorithm
had bugs and was more complex. The current algorithm
builds up the final graph by starting at the roots
with the normalized versions of anything reachable.

### CODE REVIEW

**Comment by johnynek** [comment]
cc @non

**Comment by johnynek** [comment]
@ianoc review?

**Comment by ianoc** [comment]
lgtm

**Comment by kevingessner** [comment]
@johnynek This change appears to be causing some problems in my repository -- when I run `generate`, some dependencies that are in dependencies.yaml are not generated.  Reverting this commit fixes it.

I'm trying to come up with a small test case to track down the problem.  I'll keep you posted.

**Comment by johnynek** [comment]
Okay.

That's strange because we explicitly find all the root dependencies and put in all their transitive dependencies, but maybe the root dependency checking is in correct.

(we tried this on 1 repo internally and worked as expected).

**Comment by johnynek** [comment]
cc @non

**Comment by non** [comment]
@kevingessner OK, thanks, feel free to pass along a minimized test case would be great, but also feel free to pass along a complete file if that ends up being too much work.

**Comment by kevingessner** [comment]
@johnynek It looks like roots that have no dependencies aren't added to the normalized graph.  For instance [com.maxmind.geoip](https://mvnrepository.com/artifact/com.maxmind.geoip/geoip-api/1.2.10) -- if you add that as the only entry in dependencies.yaml, nothing is generated.

**Comment by kevingessner** [comment]
cc @non ^

**Comment by johnynek** [comment]
yep....

that makes sense. Sorry.

We can add a test for this case, and make sure to explicitly addNode, not just addEdge.

Or you can take it! :)

File: src/scala/com/github/johnynek/bazel_deps/Normalizer.scala [hunk 1]
```diff
@@ -126,28 +134,34 @@ object Normalizer {
...
+          case h :: tail =>
+            val versionedH = MavenCoordinate(h, canonicals(h))
+            // we want to add this, and all the nodes this versionedH pointed to in original graph
+            val dependenciesUv: Set[UnversionedCoordinate] =
+              g.hasSource(versionedH).map(_.destination.unversioned)
+
+            val dependencies = dependenciesUv.map { uv =>
+              MavenCoordinate(uv, canonicals(uv))
+            }
+            // add the edges from versionedH -> deps
```
Comment (johnynek) [inline src/scala/com/github/johnynek/bazel_deps/Normalizer.scala]: here is the bug. We need to `addNode(versionedH)` no matter what. cc @non @kevingessner.

Would be nice to add a test to repro, then fix. Unfortunately I'm about to get on a plane and won't have time.

Comment (kevingessner) [inline src/scala/com/github/johnynek/bazel_deps/Normalizer.scala]: Here's a patch that makes the test fail: https://gist.github.com/kevingessner/ce8981aecb64dd2885d19e4a35fa366a

Comment (kevingessner) [inline src/scala/com/github/johnynek/bazel_deps/Normalizer.scala]: I'll try that fix.

Comment (non) [inline src/scala/com/github/johnynek/bazel_deps/Normalizer.scala]: I can fix this, I have a repro test now.

Comment (kevingessner) [inline src/scala/com/github/johnynek/bazel_deps/Normalizer.scala]: @johnynek @non Fix is here: https://github.com/johnynek/bazel-deps/pull/53

**Pull request merged**